File: baseline/AndroidClient/app/src/main/java/dev/telemachus/display/internet/security/AdvancedChannelSecurityGate.kt
package dev.telemachus.display.internet.security

import dev.telemachus.display.internet.InternetAudioRecordContract
import dev.telemachus.display.internet.InternetBulkRecordContract

data class AdvancedChannelOwner(
    val sessionId: String,
    val sessionEpoch: Long,
    val generation: Long,
) {
    val isValid: Boolean
        get() = sessionId.isNotBlank() && sessionEpoch > 0 && generation > 0
}

sealed interface AdvancedChannelBinding {
    val channel: SecurityChannel
    val isValid: Boolean

    data class Audio(
        val displayId: String,
        val streamId: Long,
    ) : AdvancedChannelBinding {
        override val channel = SecurityChannel.AUDIO
        override val isValid: Boolean
            get() = displayId.isNotBlank() && displayId.toByteArray(Charsets.UTF_8).size <= 128 && streamId > 0
    }

    class Bulk(transferId: ByteArray) : AdvancedChannelBinding {
        private val value = transferId.copyOf()
        val transferId: ByteArray
            get() = value.copyOf()
        override val channel = SecurityChannel.BULK
        override val isValid: Boolean
            get() = value.size in 16..64

        override fun equals(other: Any?): Boolean =
            other is Bulk && value.contentEquals(other.value)

        override fun hashCode(): Int = value.contentHashCode()
    }
}

data class AdvancedChannelAdmission(
    val id: Long,
    val owner: AdvancedChannelOwner,
    val binding: AdvancedChannelBinding,
    val plaintextBytes: Int,
)

class AdvancedChannelSecurityGate(
    initialOwner: AdvancedChannelOwner,
    private val limits: Limits = Limits.STANDARD,
) {
    data class Limits(
        val maximumAudioRecordBytes: Int,
        val maximumAudioBacklogBytes: Int,
        val maximumBulkRecordBytes: Int,
        val maximumBulkBacklogBytes: Int,
    ) {
        companion object {
            val STANDARD =
                Limits(
                    maximumAudioRecordBytes = InternetAudioRecordContract.MAXIMUM_PLAINTEXT_RECORD_BYTES,
                    maximumAudioBacklogBytes = 1024 * 1024,
                    maximumBulkRecordBytes = InternetBulkRecordContract.MAXIMUM_PLAINTEXT_RECORD_BYTES,
                    maximumBulkBacklogBytes = 4 * 1024 * 1024,
                )
        }
    }

    private var owner = initialOwner.also { require(it.isValid) { "Advanced channel owner is invalid" } }
    private var nextId = 1L
    private val admissions = mutableMapOf<Long, AdvancedChannelAdmission>()
    private val bufferedBytes = mutableMapOf<SecurityChannel, Int>()

    init {
        require(limits.maximumAudioRecordBytes > 0) { "Audio record limit must be positive" }
        require(limits.maximumAudioBacklogBytes >= limits.maximumAudioRecordBytes) {
            "Audio backlog limit must cover one record"
        }
        require(limits.maximumBulkRecordBytes > 0) { "Bulk record limit must be positive" }
        require(limits.maximumBulkBacklogBytes >= limits.maximumBulkRecordBytes) {
            "Bulk backlog limit must cover one record"
        }
    }

    @Synchronized
    fun reserve(
        payloadBytes: Int,
        binding: AdvancedChannelBinding,
        candidateOwner: AdvancedChannelOwner,
    ): AdvancedChannelAdmission {
        check(candidateOwner == owner) { "Advanced channel owner is stale" }
        require(binding.isValid) { "Advanced channel binding is invalid" }
        require(payloadBytes > 0) { "Advanced channel payload is empty" }
        val maximumRecord =
            if (binding.channel == SecurityChannel.AUDIO) limits.maximumAudioRecordBytes else limits.maximumBulkRecordBytes
        val maximumBacklog =
            if (binding.channel == SecurityChannel.AUDIO) limits.maximumAudioBacklogBytes else limits.maximumBulkBacklogBytes
        require(payloadBytes <= maximumRecord) { "Advanced channel payload exceeds $maximumRecord bytes" }
        val buffered = bufferedBytes[binding.channel] ?: 0
        check(buffered <= maximumBacklog - payloadBytes) { "Advanced channel backlog exceeds $maximumBacklog bytes" }
        check(nextId < Long.MAX_VALUE) { "Advanced channel admission sequence is exhausted" }
        val admission = AdvancedChannelAdmission(nextId++, candidateOwner, binding, payloadBytes)
        admissions[admission.id] = admission
        bufferedBytes[binding.channel] = buffered + payloadBytes
        return admission
    }

    @Synchronized
    fun finish(admission: AdvancedChannelAdmission) {
        check(admission.owner == owner) { "Advanced channel owner is stale" }
        check(admissions[admission.id] == admission) { "Advanced channel admission is unknown" }
        admissions.remove(admission.id)
        val channel = admission.binding.channel
        bufferedBytes[channel] = maxOf(0, (bufferedBytes[channel] ?: 0) - admission.plaintextBytes)
    }

    @Synchronized
    fun replaceOwner(replacement: AdvancedChannelOwner) {
        require(replacement.isValid) { "Advanced channel owner is invalid" }
        owner = replacement
        nextId = 1
        admissions.clear()
        bufferedBytes.clear()
    }

    @Synchronized
    fun bufferedBytes(channel: SecurityChannel): Int = bufferedBytes[channel] ?: 0
}

File: baseline/MacHost/Sources/Phase3/Security/AdvancedChannelSecurityGate.swift
import Foundation

struct AdvancedChannelOwner: Equatable {
    let sessionIdentifier: String
    let sessionEpoch: UInt64
    let generation: UInt64

    var isValid: Bool {
        !sessionIdentifier.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty &&
            sessionEpoch > 0 && generation > 0
    }
}

enum AdvancedChannelBinding: Equatable {
    case audio(displayID: String, streamID: UInt64)
    case bulk(transferID: Data)

    var channel: PlatformSecurityChannel {
        switch self {
        case .audio: return .audio
        case .bulk: return .bulk
        }
    }

    var isValid: Bool {
        switch self {
        case .audio(let displayID, let streamID):
            return !displayID.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty &&
                displayID.utf8.count <= 128 && streamID > 0
        case .bulk(let transferID):
            return (16...64).contains(transferID.count)
        }
    }
}

struct AdvancedChannelAdmission: Equatable {
    let identifier: UInt64
    let owner: AdvancedChannelOwner
    let binding: AdvancedChannelBinding
    let plaintextBytes: Int
}

enum AdvancedChannelSecurityError: Error, Equatable {
    case invalidOwner
    case invalidLimits
    case invalidBinding
    case emptyPayload
    case payloadTooLarge(maximum: Int)
    case backlogExceeded(maximum: Int)
    case staleOwner
    case unknownAdmission
    case sequenceExhausted
}

/// Bounded admission for Phase 5 channels. The caller must reserve before
/// sealing and finish after the corresponding DataChannel send completes.
final class AdvancedChannelSecurityGate {
    struct Limits: Equatable {
        static let standard = Limits(
            maximumAudioRecordBytes: InternetAudioRecordContract.maximumPlaintextRecordBytes,
            maximumAudioBacklogBytes: 1024 * 1_024,
            maximumBulkRecordBytes: InternetBulkRecordContract.maximumPlaintextRecordBytes,
            maximumBulkBacklogBytes: 4 * 1024 * 1024
        )

        let maximumAudioRecordBytes: Int
        let maximumAudioBacklogBytes: Int
        let maximumBulkRecordBytes: Int
        let maximumBulkBacklogBytes: Int
    }

    private let limits: Limits
    private let lock = NSLock()
    private var owner: AdvancedChannelOwner
    private var nextIdentifier: UInt64 = 1
    private var admissions: [UInt64: AdvancedChannelAdmission] = [:]
    private var bufferedBytes: [PlatformSecurityChannel: Int] = [:]

    init(owner: AdvancedChannelOwner, limits: Limits = .standard) throws {
        guard owner.isValid else {
            throw AdvancedChannelSecurityError.invalidOwner
        }
        guard limits.maximumAudioRecordBytes > 0,
              limits.maximumAudioBacklogBytes >= limits.maximumAudioRecordBytes,
              limits.maximumBulkRecordBytes > 0,
              limits.maximumBulkBacklogBytes >= limits.maximumBulkRecordBytes else {
            throw AdvancedChannelSecurityError.invalidLimits
        }
        self.owner = owner
        self.limits = limits
    }

    func reserve(
        payloadBytes: Int,
        binding: AdvancedChannelBinding,
        owner candidate: AdvancedChannelOwner
    ) throws -> AdvancedChannelAdmission {
        try synchronized {
            guard candidate == owner else { throw AdvancedChannelSecurityError.staleOwner }
            guard binding.isValid else { throw AdvancedChannelSecurityError.invalidBinding }
            guard payloadBytes > 0 else { throw AdvancedChannelSecurityError.emptyPayload }
            let channel = binding.channel
            let maximumRecord = channel == .audio
                ? limits.maximumAudioRecordBytes
                : limits.maximumBulkRecordBytes
            let maximumBacklog = channel == .audio
                ? limits.maximumAudioBacklogBytes
                : limits.maximumBulkBacklogBytes
            guard payloadBytes <= maximumRecord else {
                throw AdvancedChannelSecurityError.payloadTooLarge(maximum: maximumRecord)
            }
            let buffered = bufferedBytes[channel] ?? 0
            guard buffered <= maximumBacklog - payloadBytes else {
                throw AdvancedChannelSecurityError.backlogExceeded(maximum: maximumBacklog)
            }
            guard nextIdentifier < UInt64.max else {
                throw AdvancedChannelSecurityError.sequenceExhausted
            }
            let admission = AdvancedChannelAdmission(
                identifier: nextIdentifier,
                owner: candidate,
                binding: binding,
                plaintextBytes: payloadBytes
            )
            nextIdentifier += 1
            admissions[admission.identifier] = admission
            bufferedBytes[channel] = buffered + payloadBytes
            return admission
        }
    }

    func finish(_ admission: AdvancedChannelAdmission) throws {
        try synchronized {
            guard admission.owner == owner else { throw AdvancedChannelSecurityError.staleOwner }
            guard admissions[admission.identifier] == admission else {
                throw AdvancedChannelSecurityError.unknownAdmission
            }
            admissions.removeValue(forKey: admission.identifier)
            let channel = admission.binding.channel
            bufferedBytes[channel] = max(0, (bufferedBytes[channel] ?? 0) - admission.plaintextBytes)
        }
    }

    func replaceOwner(with replacement: AdvancedChannelOwner) throws {
        try synchronized {
            guard replacement.isValid else { throw AdvancedChannelSecurityError.invalidOwner }
            owner = replacement
            nextIdentifier = 1
            admissions.removeAll()
            bufferedBytes.removeAll()
        }
    }

    func bufferedBytes(for channel: PlatformSecurityChannel) -> Int {
        synchronized { bufferedBytes[channel] ?? 0 }
    }

    private func synchronized<T>(_ operation: () throws -> T) rethrows -> T {
        lock.lock()
        defer { lock.unlock() }
        return try operation()
    }
}
